contrib-viz/01.contrib-git.ipynb

```diff
@@ -0,0 +1,591 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Extract GitHub contributions over time\n",
+    "\n",
+    "Uses [GitPython](http://gitpython.readthedocs.io/en/stable/)."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import collections\n",
+    "import pathlib\n",
+    "import re\n",
+    "import datetime\n",
+    "import git\n",
+    "import pandas\n",
+    "import pytz"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "2c88132aa6e259244b474470bfe72722f89f1683\n"
+     ]
+    }
+   ],
+   "source": [
+    "repo = git.Repo(path = '../')\n",
+    "repo.head.reference = \"master\"\n",
+    "print(repo.head.commit)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# These patterns are reversed because our diffs are backwards\n",
+    "# -R (R=True, suggested in https://git.io/vdigv) doesn't seem to work\n",
+    "added_pattern = re.compile(r'^-(.+)', flags=re.MULTILINE)\n",
+    "deleted_pattern = re.compile(r'^\\+(.+)', flags=re.MULTILINE)\n",
+    "pattern_alphanum = re.compile(r'\\w')\n",
+    "\n",
+    "def get_word_count(text):\n",
+    "    \"\"\"\n",
+    "    Compute word count of text. Splits by whitespace and only retains\n",
+    "    words with at least one alphanumeric character [a-zA-Z0-9_].\n",
+    "    \"\"\"\n",
+    "    words = text.split()\n",
+    "    words = list(filter(pattern_alphanum.match, words))\n",
+    "    return len(words)\n",
+    "\n",
+    "def get_commit_stats(commit):\n",
+    "    \"\"\"\n",
+    "    Return addition and deletion (based on `git diff --word-diff`) for a commit.\n",
+    "    \n",
+    "    https://git-scm.com/docs/git-diff\n",
+    "    http://gitpython.readthedocs.io/en/stable/reference.html#module-git.diff\n",
+    "    \"\"\"\n",
+    "    diffs = commit.diff(\n",
+    "        other=commit.parents[0] if commit.parents else git.NULL_TREE,\n",
+    "        paths=['sections/*.md', 'content/*.md'],\n",
+    "        create_patch=True,\n",
+    "        word_diff='porcelain',\n",
+    "    )\n",
+    "\n",
+    "    delta = collections.OrderedDict()\n",
+    "    for key in 'words_added', 'words_deleted', 'characters_added', 'characters_deleted':\n",
+    "        delta[key] = 0\n",
+    "    for diff in diffs:\n",
+    "        diff_str = str(diff)\n",
+    "        additions = added_pattern.findall(diff_str)\n",
+    "        deletions = deleted_pattern.findall(diff_str)\n",
+    "\n",
+    "        # Attempt to ignore relocated lines\n",
+    "        additions, deletions = (\n",
+    "            [x for x in additions if get_word_count(x) < 4 or x not in set(deletions)],\n",
```

Comment:
Why less than 4?

Reply:
Great question! This is mostly Daniel's code. I'll play around with it and figure it out!

Reply:
That heuristic originated in https://github.com/greenelab/meta-review/commit/3aa682608566d55daecf0051af6c1764d008d3c7  Daniel's commit message acknowledges it is not very precise:
> This method aims to ignore lines that were relocated in the manuscript.
It is not very advanced and probably doesn't work very well.